File: src/engine/model/gametypes/Regicide.java
package engine.model.gametypes;

import engine.entities.Player;
import engine.model.GameType;

/**
 * A game mode where there's is a King
 * The King earn more points for a kill.
 * If a non-king kills a King he gets more points for killing the king.
 * Killing a normal player gives 1 point.
 */
public class Regicide extends GameType {
    private int kingID;
    private int scoreNeeded;
    private Player king;

       /**
     * Constructor
     *
     * @param king        the first king
     * @param scoreNeeded the score needed for a player to win
     */
    public Regicide(Player king, int scoreNeeded) {
        super(Type.Regicide);
        this.king = king;
        this.scoreNeeded = scoreNeeded;
        this.kingID = king.getId();
    }

    /**
     * Constructor
     *
     * @param kingID        the first king
     * @param scoreNeeded the score needed for a player to win
     */
    public Regicide(int kingID, int scoreNeeded) {
        super(Type.Regicide);
        this.kingID = kingID;
        this.scoreNeeded = scoreNeeded;
    }







    public int getScoreNeeded() {
        return scoreNeeded;
    }

    public Player getKing() {
        return king;
    }

    public void setKing(Player king) {
        this.king = king;
    }

    public void setKingId(int kingId) {
        this.kingID = kingId;
    }

    public int getKingId() {
        return kingID;
    }
}

File: neon-element-cpp/src/engine/model/gametypes/Regicide.cpp
#include "Regicide.hpp"
#include "engine/entities/Player.hpp"

namespace neon {

Regicide::Regicide(Player* king, int scoreNeeded)
    : GameType(Type::Regicide), king_(king), scoreNeeded_(scoreNeeded)
{
    if (king_)
        kingId_ = king_->getId();
}

Regicide::Regicide(int kingId, int scoreNeeded)
    : GameType(Type::Regicide), kingId_(kingId), scoreNeeded_(scoreNeeded)
{
}

void Regicide::setKing(Player* king)
{
    king_ = king;
    if (king_)
        kingId_ = king_->getId();
}

} // namespace neon
